keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertex.java

```diff
@@ -0,0 +1,93 @@
+package io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.unary;
+
+import io.improbable.keanu.annotation.ExportVertexToPythonBindings;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.LoadVertexParam;
+import io.improbable.keanu.vertices.SaveVertexParam;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.dbl.Differentiable;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.PartialDerivative;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+public class PermuteVertex extends DoubleUnaryOpVertex implements Differentiable {
+
+    private final static String REARRANGE_NAME = "rearrange";
+
+    private final int[] rearrange;
+
+    @ExportVertexToPythonBindings
+    public PermuteVertex(@LoadVertexParam(INPUT_VERTEX_NAME) DoubleVertex inputVertex,
+                         @LoadVertexParam(REARRANGE_NAME) int... rearrange) {
+        super(calculatePermutedShape(inputVertex, rearrange), inputVertex);
+        this.rearrange = rearrange;
+    }
+
+    @Override
+    protected DoubleTensor op(DoubleTensor value) {
+        return value.permute(rearrange);
+    }
+
+    @Override
+    public PartialDerivative forwardModeAutoDifferentiation(Map<Vertex, PartialDerivative> derivativeOfParentsWithRespectToInput) {
+        PartialDerivative derivativeOfParentWithRespectToInputs = derivativeOfParentsWithRespectToInput.get(inputVertex);
+        int[] permuteToApply = forwardPermute(derivativeOfParentWithRespectToInputs);
+        return derivativeOfParentWithRespectToInputs.permute(permuteToApply);
+    }
+
+    @Override
+    public Map<Vertex, PartialDerivative> reverseModeAutoDifferentiation(PartialDerivative derivativeOfOutputWithRespectToSelf) {
+        Map<Vertex, PartialDerivative> partials = new HashMap<>();
+        int[] permuteToApply = reversePermute(derivativeOfOutputWithRespectToSelf);
+        partials.put(inputVertex, derivativeOfOutputWithRespectToSelf.permute(permuteToApply));
+        return partials;
+    }
+
+    private int[] forwardPermute(PartialDerivative partial) {
```

Comment:
No need to iterate twice nor do you need the wrtShape.
```
    private int[] forwardPermute(PartialDerivative partial) {

        int[] permuteToApply = new int[partial.get().getRank()];

        for (int i = 0; i < partial.get().getRank(); i++) {
            if (i < rearrange.length) {
                permuteToApply[i] = rearrange[i];
            } else {
                permuteToApply[i] = i;
            }
        }

        return permuteToApply;
    }
```

Reply:
I thought splitting it into two smaller iterations would make it clearer. Yours does look a lot clearer given we don't need all of my `ofShape` and `wrtShape`!